src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -0,0 +1,297 @@
+        //
+        // For `words` = a_0 + 2^10 a_1 + ... + 2^{120} a_{12}}, initialize z_0 = `words`.
+        // If `words` fits in 130 bits, we end up with z_{13} = 0.
+        let mut z = z_0;
+        let inv_2_pow_k = F::from_u64(1u64 << K).invert().unwrap();
+        for (idx, word) in bits.iter().enumerate() {
+            let word = word
+                .clone()
+                .map(|word| F::from_u64(lebs2ip_k(&word) as u64));
+
+            // z_next = (z_cur - m_cur) / 2^K
+            z = {
+                let z_val = z
+                    .value()
+                    .zip(word)
+                    .map(|(z, word)| (z - word) * inv_2_pow_k);
+
+                // Assign z_next
+                let z_cell = region.assign_advice(
+                    || format!("z_{:?}", idx + 1),
+                    self.running_sum,
+                    offset + idx + 1,
+                    || z_val.ok_or(Error::SynthesisError),
+                )?;
+
+                CellValue::new(z_cell, z_val)
+            };
+            zs.push(z);
+        }
+
+        Ok(zs)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::{CellValue, UtilitiesInstructions, Var};
+    use super::LookupRangeCheckConfig;
+    use crate::primitives::sinsemilla::{lebs2ip_k, K};
+    use ff::PrimeFieldBits;
+    use halo2::{
+        circuit::{layouter::SingleChipLayouter, Layouter},
+        dev::MockProver,
+        plonk::{Assignment, Circuit, ConstraintSystem, Error},
+    };
+    use pasta_curves::{arithmetic::FieldExt, pallas};
+
+    use std::marker::PhantomData;
+
+    #[test]
+    fn lookup_range_check() {
+        struct MyCircuit<F: FieldExt + PrimeFieldBits> {
+            _marker: PhantomData<F>,
+        }
+
+        impl<F: FieldExt + PrimeFieldBits> UtilitiesInstructions<F> for MyCircuit<F> {
+            type Var = CellValue<F>;
+        }
+
+        impl<F: FieldExt + PrimeFieldBits> Circuit<F> for MyCircuit<F> {
+            type Config = LookupRangeCheckConfig<F, K>;
+
+            fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+                let running_sum = meta.advice_column();
+                let table_idx = meta.fixed_column();
+                let q_lookup = meta.fixed_column();
+
+                let perm = meta.permutation(&[running_sum.into()]);
+
+                LookupRangeCheckConfig::<F, K>::configure(
+                    meta,
+                    q_lookup,
+                    running_sum,
+                    table_idx,
+                    perm,
+                )
+            }
+
+            fn synthesize(
+                &self,
+                cs: &mut impl Assignment<F>,
+                config: Self::Config,
+            ) -> Result<(), Error> {
+                let mut layouter = SingleChipLayouter::new(cs)?;
+
+                // Load table_idx
+                config.load(&mut layouter)?;
+
+                let num_words = 6;
+                let words_and_expected_final_zs = [
```

Comment:
Similarly rename these throughout to refer to "elements" instead of "words" for clarity:
```suggestion
                let elements_and_expected_final_zs = [
```